File: longest-substring-without-repeating-characters/3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s==null || s.length()==0){
            return 0;
        }
        int leftIndex = 0, rightIndex = 0, length = s.length();
    // only 128 character is sc2
    boolean [] table =  new boolean[128];
    int max = 0;

    //loop through the whole string s
    while (rightIndex < length){
        if (table[s.charAt(rightIndex)]==false){
            table[s.charAt(rightIndex)] = true;
            rightIndex++;
        }else{
            max = Math.max(max,rightIndex-leftIndex);
            while (leftIndex < rightIndex && s.charAt(leftIndex) != s.charAt(rightIndex)){
                table[s.charAt(leftIndex)] = false;
                leftIndex++;
            }
            leftIndex++;
            rightIndex++;
        }
    }
    max = Math.max(max, rightIndex - leftIndex);
    return max;
    }
}

File: longest-substring-without-repeating-characters/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        b, m, d = 0, 0, {}
        # When get() is called, Python checks if the specified key exists in the dict. 
        # If it does, then get() returns the value of that key. If the key does not exist, 
        # then get() returns the value specified in the second argument to get().
        for i, l in enumerate(s): 
            #b(keft index need ++) when find again, but first one b need keep as 0 so use get(,-1)+1 
            b, m, d[l] = max(b, d.get(l, -1) + 1), max(m, i - b), i

        return max(m, len(s) - b)

